pkg/destroy/kubevirt/destroyer.go

```diff
@@ -0,0 +1,178 @@
+package kubevirt
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	apilabels "k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	ickubevirt "github.com/openshift/installer/pkg/asset/installconfig/kubevirt"
+	"github.com/openshift/installer/pkg/destroy/providers"
+	"github.com/openshift/installer/pkg/types"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// deleteFunc is the interface a function needs to implement to be delete resources.
+type deleteFunc func(namespace string, listOpts metav1.ListOptions, kubevirtClient ickubevirt.Client) error
+
+// ClusterUninstaller holds the Metadata info needed to delete the tenantCluster resources from the infraCluster.
+type ClusterUninstaller struct {
+	Metadata types.ClusterMetadata
+	Logger   logrus.FieldLogger
+}
+
+// New returns KubeVirt Uninstaller from ClusterMetadata.
+func New(logger logrus.FieldLogger, metadata *types.ClusterMetadata) (providers.Destroyer, error) {
+	return &ClusterUninstaller{
+		Metadata: *metadata,
+		Logger:   logger,
+	}, nil
+}
+
+// Run is the entrypoint to start the uninstall process.
+func (uninstaller *ClusterUninstaller) Run() error {
+	namespace := uninstaller.Metadata.Kubevirt.Namespace
+
+	listOpts := metav1.ListOptions{LabelSelector: apilabels.FormatLabels(uninstaller.Metadata.Kubevirt.Labels)}
+	kubevirtClient, err := ickubevirt.NewClient()
+	if err != nil {
+		return err
+	}
+
+	deleteFuncs := []deleteFunc{uninstaller.deleteAllVMs, uninstaller.deleteAllDVs, uninstaller.deleteAllSecrets}
+	var results = make(chan error, len(deleteFuncs))
+	for i, del := range deleteFuncs {
+		go func(index int, delFunc deleteFunc) {
```

Comment:
I am fine with doing each in its own go func, but I don't think it is necessary. You can delete sequentially, so long as each delete ends in a timely fashion and everything is retried until there are no errors.

This is one idea for how to do this.
```
  for
    for all vms
      if vm not already deleted, delete vm
    for all dvs
      if dv not already deleted, delete dv
    for all secrets
      if secret not already deleted, delete secret
    if count of vms + dvs + secrets == 0, break
    sleep
```
With this, you do not need to worry about the exponential backoff on each individual resource.

Reply:
Added jira task OCPCNV-157